File: kmp/shared/src/commonMain/kotlin/com/migestor/shared/usecase/NotebookWorkGroupPolicy.kt
package com.migestor.shared.usecase

import com.migestor.shared.domain.NotebookTab
import com.migestor.shared.domain.NotebookWorkGroup
import com.migestor.shared.domain.NotebookWorkGroupMember

/**
 * Regla única de visibilidad de grupos de trabajo.
 *
 * Los grupos con SA se ven en toda la clase. Los grupos generales se ven
 * en la pestaña de evaluación actual y en sus hijas.
 */
object NotebookWorkGroupPolicy {
    const val MODE_NONE = "none"
    const val MODE_GENERAL = "general"
    const val MODE_SITUATION_PREFIX = "situation_"

    fun canonicalTabId(
        tabs: List<NotebookTab>,
        requestedTabId: String?,
        selectedTabId: String?,
    ): String? {
        val requested = requestedTabId?.takeIf { candidate -> tabs.any { it.id == candidate } }
        val selected = selectedTabId?.takeIf { candidate -> tabs.any { it.id == candidate } }
        val seed = requested ?: selected
        if (seed != null) {
            val seedTab = tabs.first { it.id == seed }
            return seedTab.parentTabId ?: seedTab.id
        }
        val roots = tabs.filter { it.parentTabId == null }.sortedWith(compareBy<NotebookTab> { it.order }.thenBy { it.id })
        return roots.firstOrNull()?.id ?: tabs.firstOrNull()?.id
    }

    fun tabFamilyIds(tabs: List<NotebookTab>, activeTabId: String?): Set<String> {
        if (tabs.isEmpty()) return emptySet()
        val canonical = canonicalTabId(tabs, activeTabId, activeTabId) ?: return tabs.map { it.id }.toSet()
        return buildSet {
            add(canonical)
            tabs.filter { it.parentTabId == canonical }.forEach { add(it.id) }
            tabs.firstOrNull { it.id == activeTabId }?.parentTabId?.let { parent ->
                add(parent)
                tabs.filter { it.parentTabId == parent }.forEach { add(it.id) }
            }
        }
    }

    fun groupsForManagement(
        groups: List<NotebookWorkGroup>,
        tabs: List<NotebookTab>,
        selectedTabId: String?,
    ): List<NotebookWorkGroup> {
        val family = tabFamilyIds(tabs, selectedTabId)
        val inFamily = groups.filter { family.isEmpty() || it.tabId in family }
        val saGroups = groups.filter { it.learningSituationId != null }
        return (inFamily + saGroups)
            .distinctBy { it.id }
            .sortedWith(compareBy<NotebookWorkGroup> { it.order }.thenBy { it.id })
    }

    fun activeGroups(
        groups: List<NotebookWorkGroup>,
        members: List<NotebookWorkGroupMember>,
        tabs: List<NotebookTab>,
        activeTabId: String?,
        mode: String,
    ): List<NotebookWorkGroup> {
        if (mode == MODE_NONE) return emptyList()
        val family = tabFamilyIds(tabs, activeTabId)
        val inFamily = groups.filter { family.isEmpty() || it.tabId in family }
        val hasMembers: (NotebookWorkGroup) -> Boolean = { group ->
            members.any { it.groupId == group.id }
        }

        val selected = if (mode.startsWith(MODE_SITUATION_PREFIX)) {
            val sitId = mode.removePrefix(MODE_SITUATION_PREFIX).toLongOrNull()
            val sitGroups = groups.filter { it.learningSituationId == sitId }
            sitGroups.ifEmpty { inFamily }
        } else {
            val candidates = inFamily.ifEmpty { groups }
            val filled = candidates.filter(hasMembers)
            filled.ifEmpty { candidates }
        }

        return selected.sortedWith(compareBy<NotebookWorkGroup> { it.order }.thenBy { it.id })
    }

    fun memberIds(
        group: NotebookWorkGroup,
        members: List<NotebookWorkGroupMember>,
    ): Set<Long> {
        return members.filter { it.groupId == group.id }.map { it.studentId }.toSet()
    }

    fun groupIdForStudent(
        studentId: Long,
        members: List<NotebookWorkGroupMember>,
        activeGroups: List<NotebookWorkGroup>,
    ): Long? {
        val activeIds = activeGroups.map { it.id }.toSet()
        return members.firstOrNull { member ->
            member.studentId == studentId && (activeIds.isEmpty() || member.groupId in activeIds)
        }?.groupId ?: members.firstOrNull { it.studentId == studentId }?.groupId
    }
}

File: kmp/iosApp/AppleShared/NotebookWorkGroupPolicy.swift
import Foundation
import MiGestorKit

enum NotebookWorkGroupPolicy {
    static let modeNone = "none"
    static let modeGeneral = "general"
    static let situationPrefix = "situation_"

    static func canonicalTabId(
        tabs: [NotebookTab],
        requestedTabId: String?,
        selectedTabId: String?
    ) -> String? {
        let requested = requestedTabId.flatMap { candidate in
            tabs.contains(where: { $0.id == candidate }) ? candidate : nil
        }
        let selected = selectedTabId.flatMap { candidate in
            tabs.contains(where: { $0.id == candidate }) ? candidate : nil
        }
        if let seed = requested ?? selected, let seedTab = tabs.first(where: { $0.id == seed }) {
            return seedTab.parentTabId ?? seedTab.id
        }
        let roots = tabs.filter { $0.parentTabId == nil }.sorted {
            if $0.order != $1.order { return $0.order < $1.order }
            return $0.id < $1.id
        }
        return roots.first?.id ?? tabs.first?.id
    }

    static func tabFamilyIds(tabs: [NotebookTab], activeTabId: String?) -> Set<String> {
        guard !tabs.isEmpty else { return [] }
        guard let canonical = canonicalTabId(tabs: tabs, requestedTabId: activeTabId, selectedTabId: activeTabId) else {
            return Set(tabs.map(\.id))
        }
        var family: Set<String> = [canonical]
        tabs.filter { $0.parentTabId == canonical }.forEach { family.insert($0.id) }
        if let selected = tabs.first(where: { $0.id == activeTabId }), let parent = selected.parentTabId {
            family.insert(parent)
            tabs.filter { $0.parentTabId == parent }.forEach { family.insert($0.id) }
        }
        return family
    }

    static func groupsForManagement(
        groups: [NotebookWorkGroup],
        tabs: [NotebookTab],
        selectedTabId: String?
    ) -> [NotebookWorkGroup] {
        let family = tabFamilyIds(tabs: tabs, activeTabId: selectedTabId)
        let inFamily = groups.filter { family.isEmpty || family.contains($0.tabId) }
        let saGroups = groups.filter { $0.learningSituationId != nil }
        var seen = Set<Int64>()
        return (inFamily + saGroups).filter { group in
            seen.insert(group.id).inserted
        }.sorted {
            if $0.order != $1.order { return $0.order < $1.order }
            return $0.id < $1.id
        }
    }

    static func activeGroups(
        groups: [NotebookWorkGroup],
        members: [NotebookWorkGroupMember],
        tabs: [NotebookTab],
        activeTabId: String?,
        mode: String
    ) -> [NotebookWorkGroup] {
        guard mode != modeNone else { return [] }
        let family = tabFamilyIds(tabs: tabs, activeTabId: activeTabId)
        let inFamily = groups.filter { family.isEmpty || family.contains($0.tabId) }
        let hasMembers: (NotebookWorkGroup) -> Bool = { group in
            members.contains { $0.groupId == group.id }
        }

        let selected: [NotebookWorkGroup]
        if mode.hasPrefix(situationPrefix),
           let sitId = Int64(mode.dropFirst(situationPrefix.count)) {
            let sitGroups = groups.filter { $0.learningSituationId?.int64Value == sitId }
            selected = sitGroups.isEmpty ? inFamily : sitGroups
        } else {
            let candidates = inFamily.isEmpty ? groups : inFamily
            let filled = candidates.filter(hasMembers)
            selected = filled.isEmpty ? candidates : filled
        }

        return selected.sorted {
            if $0.order != $1.order { return $0.order < $1.order }
            return $0.id < $1.id
        }
    }

    static func memberIds(
        group: NotebookWorkGroup,
        members: [NotebookWorkGroupMember]
    ) -> Set<Int64> {
        Set(members.filter { $0.groupId == group.id }.map(\.studentId))
    }

    static func groupIdForStudent(
        studentId: Int64,
        members: [NotebookWorkGroupMember],
        activeGroups: [NotebookWorkGroup]
    ) -> Int64? {
        let activeIds = Set(activeGroups.map(\.id))
        return members.first(where: { member in
            member.studentId == studentId && (activeIds.isEmpty || activeIds.contains(member.groupId))
        })?.groupId ?? members.first(where: { $0.studentId == studentId })?.groupId
    }
}
